lib/mongo/socket.rb

```diff
@@ -169,7 +169,22 @@ def eof?
     private
 
     def read_from_socket(length)
-      @socket.read(length) || String.new
+      # Block on data to read for timeout seconds
+      # using the suggested implementation of http://www.ruby-doc.org/core-2.1.3/Kernel.html#method-i-select
+      # to work with SSL connections and some OSs like Ubuntu, CentOS and Red Hat
+      time_left = timeout.to_f
+      data = ""
+      begin
+        while data.length < length
+          data << @socket.read_nonblock(length - data.length)
+        end
+      rescue IO::WaitReadable
+        raise Error::OperationFailure.new("Took more than #{timeout} seconds to receive data.") if (time_left -= 0.1) <= 0
```

Comment:
We use the `OperationFailure` error specifically for errors that come from the server, for example write concern errors or query syntax failures to name a few - `TimeoutError` is outside of this classification but won't require a new error to be created.